modules/analysis/src/service.rs

```diff
@@ -0,0 +1,1024 @@
+};
+use crate::Error;
+use itertools::Itertools;
+use petgraph::graph::{Graph, NodeIndex};
+use petgraph::Direction;
+use sea_query::{Condition, Order};
+use std::str::FromStr;
+use trustify_common::db::ConnectionOrTransaction;
+use trustify_common::purl::Purl;
+use trustify_entity::relationship::Relationship;
+use trustify_entity::{sbom, sbom_node};
+
+pub struct AnalysisService {
+    db: Database,
+}
+
+pub fn dep_nodes(
+    graph: &petgraph::Graph<PackageNode, Relationship, petgraph::Directed>,
+    node: NodeIndex,
+) -> Vec<DepNode> {
+    // fast enough to directly instantiate DepNode in this function
+    let mut depnodes: Vec<DepNode> = Vec::new();
+    fn dfs(
+        graph: &petgraph::Graph<PackageNode, Relationship, petgraph::Directed>,
+        node: NodeIndex,
+        depnodes: &mut Vec<DepNode>,
+    ) {
+        for neighbor in graph.neighbors_directed(node, Direction::Incoming) {
+            if let Some(dep_packagenode) = graph.node_weight(neighbor).cloned() {
+                let dep_node = DepNode::new(
+                    dep_packagenode.purl.to_string(),
+                    dep_packagenode.name.to_string(),
+                    dep_nodes(graph, neighbor),
+                );
+
+                if !depnodes.contains(&dep_node) {
+                    depnodes.push(dep_node);
+                }
+            } else {
+                log::warn!("Node weight for neighbor {:?} not found", neighbor);
+            }
+        }
+    }
+    dfs(graph, node, &mut depnodes);
+    depnodes
+}
+
+pub fn ancestor_nodes(
+    graph: &petgraph::Graph<PackageNode, Relationship, petgraph::Directed>,
+    node: NodeIndex,
+    original_node: NodeIndex,
+) -> Vec<NodeIndex> {
+    // working directly with NodeIndex is a bit faster (we instantiate PackageNode later)
+    let mut ancestor_nodes: Vec<NodeIndex> = Vec::new();
+    if node != original_node {
+        ancestor_nodes.push(node);
+    }
+    fn dfs(
+        graph: &petgraph::Graph<PackageNode, Relationship, petgraph::Directed>,
+        node: NodeIndex,
+        original_node: NodeIndex,
+        ancestor_nodes: &mut Vec<NodeIndex>,
+    ) {
+        for neighbor in graph.neighbors_directed(node, Direction::Outgoing) {
+            if neighbor == original_node {
+                break; // no need to descend further past the original node in the tree
+            }
+            if !ancestor_nodes.contains(&neighbor) {
+                ancestor_nodes.push(neighbor);
+                dfs(graph, neighbor, original_node, ancestor_nodes);
+            }
+        }
+    }
+    dfs(graph, node, original_node, &mut ancestor_nodes);
+    ancestor_nodes
+}
+
+pub async fn get_relationships(
+    connection: &ConnectionOrTransaction<'_>,
+    distinct_sbom_id: &str,
+) -> Result<Vec<QueryResult>, DbErr> {
+    // TODO: We may convert this to sea_orm at some point though keeping this in its current form, to highlight
```

Comment:
Fair. The upside to converting to SeaORM at some point would be for refactoring and type-safety or whatnot, if we change the schema/entity models.

Non-blocking.